src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -258,6 +258,11 @@ public Node start() {
         injector.getInstance(SearchService.class).start();
         injector.getInstance(MonitorService.class).start();
         injector.getInstance(RestController.class).start();
+
+        // TODO hack around circular dependecncies problems
+        injector.getInstance(AsyncShardFetch.Started.class).setReallocation(injector.getInstance(ClusterService.class), injector.getInstance(AllocationService.class));
```

Comment:
this is super ugly I think `AsyncShardFetch.Started` and `AsyncShardFetch.Store` should be an impl detail of `GatewayAllocator` no need to bind this or anything